From patchwork Tue Sep  6 19:48:55 2022
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Liam Howlett <Liam.Howlett@Oracle.com>
X-Patchwork-Id: 12968087
Return-Path: <owner-linux-mm@kvack.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.lore.kernel.org (Postfix) with ESMTP id D5DEDC38145
	for <linux-mm@archiver.kernel.org>; Tue,  6 Sep 2022 19:49:55 +0000 (UTC)
Received: by kanga.kvack.org (Postfix)
	id 7BF338000D; Tue,  6 Sep 2022 15:49:49 -0400 (EDT)
Received: by kanga.kvack.org (Postfix, from userid 40)
	id 73E708000F; Tue,  6 Sep 2022 15:49:49 -0400 (EDT)
X-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id 100198000D; Tue,  6 Sep 2022 15:49:49 -0400 (EDT)
X-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (smtprelay0011.hostedemail.com
 [216.40.44.11])
	by kanga.kvack.org (Postfix) with ESMTP id C5F378000D
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 15:49:48 -0400 (EDT)
Received: from smtpin05.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay07.hostedemail.com (Postfix) with ESMTP id A8C4616084C
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:48 +0000 (UTC)
X-FDA: 79882700856.05.D250996
Received: from mx0a-00069f02.pphosted.com (mx0a-00069f02.pphosted.com
 [205.220.165.32])
	by imf07.hostedemail.com (Postfix) with ESMTP id 2CBFF40059
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:47 +0000 (UTC)
Received: from pps.filterd (m0246629.ppops.net [127.0.0.1])
	by mx0b-00069f02.pphosted.com (8.17.1.5/8.17.1.5) with ESMTP id
 286Id90r024660;
	Tue, 6 Sep 2022 19:49:43 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : references : in-reply-to : content-type :
 content-transfer-encoding : mime-version; s=corp-2022-7-12;
 bh=x1PG8mwAB7xd3Yjc+aslFwuy5lw4nLRy3s499xqRwJE=;
 b=xFpMCvStjnShctz9EX3mj61WS2cXBmdYOlFaqgPGD1ItvqMBw7YB74wAbTnM+JkiM5rX
 r1zinHNU20GkEEv79kkT/TPSss12prpbmwgxn8Sqz2SpdF1ftzUd9gddzeCTrL5m07cM
 BhhMcAuZP/cJrymEPFJwieh1ix46W6SxwKsg7EptFpijlzItL/v7ZXUnlHQR2QaojU9e
 EBEYeOH870Q5b8gxWs4f1rCCmo1c1vBaxL5hMZBN1gr8nA+zyKv9zbHHZMtj9qKgsZIg
 SVaE+x7mqZ+W+JXnC/d5DDwjfcEQAtk7ExyZwEIt8//F2/opl1X7BS2iptHAXeQnBAng bg==
Received: from iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com
 (iadpaimrmta03.appoci.oracle.com [130.35.103.27])
	by mx0b-00069f02.pphosted.com (PPS) with ESMTPS id 3jbxtaet7s-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:49:43 +0000
Received: from pps.filterd
 (iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com [127.0.0.1])
	by iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com (8.17.1.5/8.17.1.5)
 with ESMTP id 286Je5WS039711;
	Tue, 6 Sep 2022 19:49:41 GMT
Received: from nam10-dm6-obe.outbound.protection.outlook.com
 (mail-dm6nam10lp2108.outbound.protection.outlook.com [104.47.58.108])
	by iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com (PPS) with ESMTPS id
 3jbwc9m7j9-5
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:49:41 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=mvF4D1xS8qAciSD2t7RY2p7x9a72ByhLTjbHcqMvKjojCULAd9EGQ06XN6IY3vBAbbBCjzK285GzMiKw6aUzLBkvtdjsENRIpn2p85G8O9fC2qPj3szzrOsY9ms9Iuy+UDxa/VirUD1P9NS/zp21x45EM7I85PRlfaUfhbjvG30a7C9o/1/Zcosj6zvO/q7ELvrMLPJjmAnBgLpFoHKymstJ2mDqq3RZDARd6bvGYbQffAC0/6eEeX42MPT6o9sFbGD6eH8TOV9CKNgZVieQe76B88U8LlOGrgkwYAIbNPfcVKXPewf3zbySMxNJZ7mynvmHEecnHmwRVidSj0sPMw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-AntiSpam-MessageData-ChunkCount:X-MS-Exchange-AntiSpam-MessageData-0:X-MS-Exchange-AntiSpam-MessageData-1;
 bh=x1PG8mwAB7xd3Yjc+aslFwuy5lw4nLRy3s499xqRwJE=;
 b=Hf7YrLMme0NzdcQ/G2VQsesFu2oyJMsCAR7C3omKvK9mqd8yO5bq4vPgi58888CaxviIg/uc1DTb0yt0sm3ZBZriz9V7KuvouCw3sMCoigJF8TDslliljSRDWMpSHcab1QQgTb3k1iKDu/qBjRxSaskWLslnXfojDTAa4wZvacrgcmLAptQwnIjwbiUtykwBJScWqS7vqVYp+Akgns4KlWxTt52pQ4qGMa79jsQVJRxHknK4Cft/FhEorFYRYaSp0YdcvecGsH1JRiCJxcLUQBvHgqIU2lRchC9wB3YaIs507OQ7yUcxSi94NOSGHV/RlC9f2TD0fWkSABTD6SG9gQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=x1PG8mwAB7xd3Yjc+aslFwuy5lw4nLRy3s499xqRwJE=;
 b=uzF6KBibaLIynso4vafLd/ODKllaB2MSRfjCmz1nFElmuNDFlgkAOoQ/GaRRP3L+fP3ceIypNovPp0EprvPLnyGVdgw1PtQAquB2FLl6hh/LB/hsDOpqUpaM3/X5TQVYt3mvzHiWpcNvcvA1GuHfs125bLATv7BHFiVytyxJ4bY=
Received: from SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25)
 by SJ0PR10MB5598.namprd10.prod.outlook.com (2603:10b6:a03:3d9::18) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.5588.10; Tue, 6 Sep
 2022 19:49:38 +0000
Received: from SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223]) by SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223%5]) with mapi id 15.20.5588.018; Tue, 6 Sep 2022
 19:49:38 +0000
From: Liam Howlett <liam.howlett@oracle.com>
To: "maple-tree@lists.infradead.org" <maple-tree@lists.infradead.org>,
        "linux-mm@kvack.org" <linux-mm@kvack.org>,
        "linux-kernel@vger.kernel.org"
	<linux-kernel@vger.kernel.org>,
        Andrew Morton <akpm@linux-foundation.org>
CC: "Matthew Wilcox (Oracle)" <willy@infradead.org>,
        Liam Howlett
	<liam.howlett@oracle.com>,
        Vlastimil Babka <vbabka@suse.cz>, Davidlohr Bueso
	<dave@stgolabs.net>
Subject: [PATCH v14 38/70] optee: remove vma linked list walk
Thread-Topic: [PATCH v14 38/70] optee: remove vma linked list walk
Thread-Index: AQHYwimybpkPdOjfJEiVWqcHTsa6PA==
Date: Tue, 6 Sep 2022 19:48:55 +0000
Message-ID: <20220906194824.2110408-39-Liam.Howlett@oracle.com>
References: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
In-Reply-To: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-mailer: git-send-email 2.35.1
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 91e7e59a-8a7b-427f-9741-08da9040eefe
x-ms-traffictypediagnostic: SJ0PR10MB5598:EE_
x-ms-exchange-senderadcheck: 1
x-ms-exchange-antispam-relay: 0
x-microsoft-antispam: BCL:0;
x-microsoft-antispam-message-info: 
 dYiElD4FkGaQQqilVBZ03hmkkOVA4024esMkjRBjD6FYLJ4/YsnoMUUjoh/SU5f6DlVM7rWpOKUVhc0AvEgsYiNg7bRmMJn3RDQKFUl3f9YQKVR8afdnYjs6E1A2FMZ7DK4mRpOYcuiTNtHjUKoVNS5J7EeBjy/MCccCol6rbCHu/E+NFWXzcNjMpZaASKzLT1I/0nNkZf+Z1izqgKRxYiAGfDvmflhGkjC7Tn2MxoB3Q3fzkvb7Anfe8T9ugOziGE4Klr6WbR2MUFX60jZ94XFhWGpbYC6lGNsrO9p7PYV3Bxoduyi2alUIc/bM1ilVqWpTNRrM71AXD0IyrNdWYMLE2S8AbnMqORhEEdU0DLITNsfyo9FvP+arxlAUL0LgIEjXiEvBCws63s3PXEyEmBSwgzqHgbVgCeJN6JI6mbz9rBeiGmGftykwMqZaMZoz+g1klxE91qWme7hp6nAhlqoQE9gQEk4zEr79v2jZZVfwqdwWmHR2Ze0dt7cmXvPFGzIuxcLQ0xPhtV/Ab/3cx1DTWDZgCUqcAnrd3TclZaVxR9BpGADON5ehwdp9RtqFOodaR0L8M9I/qG9IOyOPa1yFlHUkVaW2ZQGEurXHN8PwxpDzlGLafrRiudHryoC5Mmg4U4CKn56i8us0SRN3z5s9P0YasB+CctlWP8E0BKIYMbMz/ey1G9iPnW6ww+sJnLF+eiP6iQN1EWeXpjRtB+H7HRYnrRDjkkU46bHAHoDGW3i+ftHD1AZ4ocSEmUwAdocZ3zzrp/Od4XDxlRo/Aw==
x-forefront-antispam-report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3022.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(13230016)(346002)(39860400002)(396003)(136003)(376002)(366004)(64756008)(66556008)(41300700001)(91956017)(76116006)(66476007)(66446008)(66946007)(71200400001)(54906003)(110136005)(36756003)(316002)(6512007)(44832011)(8676002)(122000001)(2906002)(6486002)(4326008)(186003)(8936002)(5660300002)(478600001)(38070700005)(38100700002)(2616005)(86362001)(26005)(1076003)(6666004)(6506007)(83380400001);DIR:OUT;SFP:1101;
x-ms-exchange-antispam-messagedata-chunkcount: 1
x-ms-exchange-antispam-messagedata-0: =?iso-8859-1?q?aSFaEIvvCv125qyojZwH/4v?=
	=?iso-8859-1?q?L2+2R7KAn3KiJEos8recbzkzT71OHV9g9Cr/F3OfvDO3Ppr4wpuizb220g13?=
	=?iso-8859-1?q?pFMoVchMwQruL+RB7YaYlIBp4d6piA+RNvH/ya6jxfKykO6Cdh3AT5b1Sa2L?=
	=?iso-8859-1?q?4sLckzN+JJDArnrSaaI138/rs+MKfNEX05N9ncSImPlxn8tg6bxf9j+de1aR?=
	=?iso-8859-1?q?AN5cugGOMGvVasSFwnCcXS+lCcabxjUM4o7KCkVKXcSvd+32e4wHeXEb4abN?=
	=?iso-8859-1?q?luQzpuYIw9qb6oB5Esvo3HCxC72LTwX0EFDdOtgPa4raH3kIFWkpw9jSHEc6?=
	=?iso-8859-1?q?aYcvUhr0NqrEernvxC1+GF3g2YUT71nDeDUfIfJtDMAdpukAV42S/ESNca19?=
	=?iso-8859-1?q?rzlPBcdWrXhP+F8v5AJafOP0ZsIhjKEyd7xDmwe7mZ2IuZrSKKrGKViEP8RG?=
	=?iso-8859-1?q?JJ6w8mbIgbTyLpFosDiilyMqLu9L9bZLTFDgm4xDjLtoK524tEvm7ZwAhgz0?=
	=?iso-8859-1?q?34x6lq8Yen5cFXePyUqPakY1D/U+9fYQBJIsmmGpA5mcmzwhrxOd/nIIiKT+?=
	=?iso-8859-1?q?98WhxnoFuBPyMTv4zmXAapwk/PjAHec0p/kM/j5E6LlJ95bAvZ9117wB0A8/?=
	=?iso-8859-1?q?VZzjnIExID2brbGGegiNAgbgrJNH6zfakrsHAnt0+cgMGznV08wK0yeM1AQl?=
	=?iso-8859-1?q?hbI1XvI8fOV0CVJt6hGt8RYWMzG3FmMHoI5vNQyVggc7i9O6CiombMCkCKUj?=
	=?iso-8859-1?q?wfjL2iq8UxH0TzwxYmgbfX5wM3tE1xYHLtJPtHQAIQFrYBeEFz0MUQ/FmTwl?=
	=?iso-8859-1?q?g0//eZ6mB7CpJCKt9UG8eqlEhi3pWyTc9okGu2jKXYG+qO/ZXqNUnbzDZwLI?=
	=?iso-8859-1?q?QZo8lhR87+m9FN+qHvWE03J19Qn3GiLHyxkmVsHHkUy6MqQDMmZ/MngH+3Hj?=
	=?iso-8859-1?q?PK+4On3/l0xaI+NyXodJ5YS7rGlYY7r4GEB+8Ui0pXWCercRmAJh/BFMaJUu?=
	=?iso-8859-1?q?roLwAuDjJMr+baTWdwpy/9NOHxbzgARZ813jeNh70NETfZi8FreaQMX7HTlw?=
	=?iso-8859-1?q?k3/k8EScOzOBiSgrLIuc+kAgZ+/BREX0wxKr5ZG2Jlp5mqgKHVg8vb2MKmj1?=
	=?iso-8859-1?q?XyHgkzWzx9f+xzq8xAAnFcv/vpkduw8fwIEPh8jmue1r+XnOc1rcb2vjFqvi?=
	=?iso-8859-1?q?4KRJDoin4RK9bASerSAW6JtP2NDk5IFugGNIMYojxXaq/JqfzWziksgqy/M7?=
	=?iso-8859-1?q?9S4JbJBntcQRsp9xh2UO5WFMSyoFnaWdOLJE18ZUkhImr+qC+2o/vJnxMJpJ?=
	=?iso-8859-1?q?F+eNbh393tUmwfZIy6osTASTsIYg2mEaC5vnVa/uf4VmuQqflzy+NusS66kg?=
	=?iso-8859-1?q?fcHNB5vO9HZzIhCCUD1JbOFYFmYtrqUrhdv1vAsonVtSYbTbRhP8QfKPDuOR?=
	=?iso-8859-1?q?abE0DmICX+0MPedmkuCFeTU2INcALMtNNr2BDIEQ1FkbECUKX/CnQPZ10fmy?=
	=?iso-8859-1?q?uFOnxekmri+QTncHi/HDLI/SoUdoqqaunb4BuUCqRfVHCf1O/68inGuO4ErR?=
	=?iso-8859-1?q?ZUBnpNlyINmoeN5eBP9xwCRRUpv/EnSYRL8C8xzUDyyZKgeOskSJImTD3NIe?=
	=?iso-8859-1?q?/CSsMd7/j/+YplkKVHX1KetWj24t9BLdZBqOI6Q=3D=3D?=
MIME-Version: 1.0
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3022.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 91e7e59a-8a7b-427f-9741-08da9040eefe
X-MS-Exchange-CrossTenant-originalarrivaltime: 06 Sep 2022 19:48:55.8500
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 gnaOgI3qE3yFIZO1iZlttikNRA38HwP2InhmcCVWlUkGvY91RAG+FeC2d7edY5eyYkxS76n0rmAGB2CS8ls5MQ==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SJ0PR10MB5598
X-Proofpoint-Virus-Version: vendor=baseguard
 engine=ICAP:2.0.205,Aquarius:18.0.895,Hydra:6.0.528,FMLib:17.11.122.1
 definitions=2022-09-06_09,2022-09-06_02,2022-06-22_01
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 bulkscore=0
 mlxlogscore=999
 phishscore=0 mlxscore=0 spamscore=0 malwarescore=0 suspectscore=0
 adultscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2207270000 definitions=main-2209060091
X-Proofpoint-ORIG-GUID: V3k-lmfXXmRBWoLuKTsPDXtd0pN5aJAO
X-Proofpoint-GUID: V3k-lmfXXmRBWoLuKTsPDXtd0pN5aJAO
ARC-Seal: i=2; s=arc-20220608; d=hostedemail.com; t=1662493788; a=rsa-sha256;
	cv=pass;
	b=vsqMAV7BxVPxjrPPuvLoBnRJlZkwJcWumN593FnExljyqWRHTsIGWNBczPHYmgQkXwsw2B
	h9yivN0w0XVejfkpS/uITdBNs2RpZFKfbmB6QLm7L0R8Iv+jLna+SZza+AYpuUHTUH3GhG
	2N6ePSF26ftiaietomRiIK3Pj8bRhj8=
ARC-Authentication-Results: i=2;
	imf07.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=xFpMCvSt;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=uzF6KBib;
	spf=pass (imf07.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
ARC-Message-Signature: i=2; a=rsa-sha256; c=relaxed/relaxed;
 d=hostedemail.com;
	s=arc-20220608; t=1662493788;
	h=from:from:sender:reply-to:subject:subject:date:date:
	 message-id:message-id:to:to:cc:cc:mime-version:mime-version:
	 content-type:content-type:
	 content-transfer-encoding:content-transfer-encoding:
	 in-reply-to:in-reply-to:references:references:dkim-signature;
	bh=x1PG8mwAB7xd3Yjc+aslFwuy5lw4nLRy3s499xqRwJE=;
	b=ho7TMFirqoLibFsTUqnuDSSpWqriKlJTFCXUxmoyt6xc+iEsMSfP2dewXqvkJRnJhNMcTS
	YYkJg33kw2foMNhp8FjaNYRxudufbS28G2LA2LAdBjZIAenCC9BsupyqWVI+D6nEwf8/2h
	Ozg0ZuZ7yNa5AXHKQBvgX/XTDoTKXQs=
X-Stat-Signature: z41d8yh43j6q8k6491896tjddxdu8scm
X-Rspamd-Queue-Id: 2CBFF40059
X-Rspamd-Server: rspam07
X-Rspam-User: 
Authentication-Results: imf07.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=xFpMCvSt;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=uzF6KBib;
	spf=pass (imf07.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
X-HE-Tag: 1662493787-401647
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

From: "Matthew Wilcox (Oracle)" <willy@infradead.org>

Use the VMA iterator instead.  Change the calling convention of
__check_mem_type() to pass in the mm instead of the first vma in the
range.

Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
Signed-off-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
Acked-by: Vlastimil Babka <vbabka@suse.cz>
Reviewed-by: Davidlohr Bueso <dave@stgolabs.net>
---
 drivers/tee/optee/call.c | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/drivers/tee/optee/call.c b/drivers/tee/optee/call.c
index 28f87cd8b3ed..290b1bb0e9cd 100644
--- a/drivers/tee/optee/call.c
+++ b/drivers/tee/optee/call.c
@@ -492,15 +492,18 @@ static bool is_normal_memory(pgprot_t p)
 #endif
 }
 
-static int __check_mem_type(struct vm_area_struct *vma, unsigned long end)
+static int __check_mem_type(struct mm_struct *mm, unsigned long start,
+				unsigned long end)
 {
-	while (vma && is_normal_memory(vma->vm_page_prot)) {
-		if (vma->vm_end >= end)
-			return 0;
-		vma = vma->vm_next;
+	struct vm_area_struct *vma;
+	VMA_ITERATOR(vmi, mm, start);
+
+	for_each_vma_range(vmi, vma, end) {
+		if (!is_normal_memory(vma->vm_page_prot))
+			return -EINVAL;
 	}
 
-	return -EINVAL;
+	return 0;
 }
 
 int optee_check_mem_type(unsigned long start, size_t num_pages)
@@ -516,8 +519,7 @@ int optee_check_mem_type(unsigned long start, size_t num_pages)
 		return 0;
 
 	mmap_read_lock(mm);
-	rc = __check_mem_type(find_vma(mm, start),
-			      start + num_pages * PAGE_SIZE);
+	rc = __check_mem_type(mm, start, start + num_pages * PAGE_SIZE);
 	mmap_read_unlock(mm);
 
 	return rc;

